File: pong_server/src/game_server/Segment2.py
from typing import Optional

import numpy


class Segment2(object):
    def __init__(self,
                 begin: numpy.ndarray,
                 end: numpy.ndarray,
                 vector: Optional[numpy.ndarray] = None):
        self.begin = numpy.array([begin[0], begin[1]], dtype=float)
        self.end = numpy.array([end[0], end[1]], dtype=float)
        if vector is not None:
            self.vector = numpy.array([vector[0], vector[1]], dtype=float)
        else:
            self.vector = self.end - self.begin

    def intersect(self, segment2) -> (Optional[numpy.ndarray], Optional[float]):
        s2 = segment2

        t1_top = ((s2.end[0] - s2.begin[0])
                  * (self.begin[1] - s2.begin[1])
                  - (s2.end[1] - s2.begin[1]) * (self.begin[0] - s2.begin[0]))

        t2_top = ((s2.begin[1] - self.begin[1])
                  * (self.begin[0] - self.end[0])
                  - (s2.begin[0] - self.begin[0]) * (self.begin[1] - self.end[1]))

        bottom = ((s2.end[1] - s2.begin[1])
                  * (self.end[0] - self.begin[0])
                  - (s2.end[0] - s2.begin[0]) * (self.end[1] - self.begin[1]))

        if bottom == 0:
            return None, None

        t1 = t1_top / bottom
        if t1 < 0 or t1 > 1:
            return None, None

        t2 = t2_top / bottom
        if t2 < 0 or t2 > 1:
            return None, None

        return (
            numpy.array([self.begin[0] + t1 * (self.end[0] - self.begin[0]),
                         self.begin[1] + t1 * (self.end[1] - self.begin[1])],
                        dtype=float),
            t1
        )

File: front/app/src/js/components/game/Scene/Segment2.js
import * as THREE from 'three';

export class Segment2 {
  #begin = new THREE.Vector2;
  #end = new THREE.Vector2;

  constructor(begin, end) {
    this.#begin.set(begin.x, begin.y);
    this.#end.set(end.x, end.y);
  }

  intersect(segment2) {
    const t1Top = (segment2.end.x - segment2.begin.x) *
            (this.#begin.y - segment2.begin.y) -
            (segment2.end.y - segment2.begin.y) *
                (this.#begin.x - segment2.begin.x);
    const t2Top = (segment2.begin.y - this.#begin.y) *
            (this.#begin.x - this.#end.x) -
            (segment2.begin.x - this.#begin.x) *
                (this.#begin.y - this.#end.y);
    const bottom = (segment2.end.y - segment2.begin.y) *
            (this.#end.x - this.#begin.x) -
            (segment2.end.x - segment2.begin.x) *
                (this.#end.y - this.#begin.y);

    if (bottom === 0) {
      return {point: null, t: null};
    }

    const t1 = t1Top / bottom;
    if (t1 < 0 || t1 > 1) {
      return {point: null, t: null};
    }

    const t2 = t2Top / bottom;
    if (t2 < 0 || t2 > 1) {
      return {point: null, t: null};
    }

    return {
      point: new THREE.Vector2(this.#begin.x + t1 *
          (this.#end.x - this.#begin.x),
      this.#begin.y + t1 * (this.#end.y - this.#begin.y)),
      t: t1,
    };
  }

  get begin() {
    return this.#begin;
  }

  get end() {
    return this.#end;
  }
}
